### PR TITLE
Add constraint values for the Rust release channel.

### PR DESCRIPTION
**Pull request by jmillikin**
This allows users to define separate toolchains and platforms for stable/beta/nightly, and to `select()` based on which channel is being used in a build.

In the upcoming Bazel v4.0, the `target_compatible_with` target attr will allow targets to be skipped based on channel compatibility.

### CODE REVIEW

**Review by mfarrugi — APPROVED** [review]

**Comment by mfarrugi** [comment]
This could use an example. Is CI setup to run 4.0 yet?

File: rust/platform/channel/BUILD [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package(default_visibility = ["//visibility:public"])
+
+# Allows toolchains and targets to constrain themselves based on Rust release
+# channels, for example when depending on features only available in Nightly.
+#
+# https://doc.rust-lang.org/book/appendix-07-nightly-rust.html
+constraint_setting(name = "channel")
```
Comment (mfarrugi) [inline rust/platform/channel/BUILD:7]: I assume this is not a global name, right? I am not familiar w/ defining constraints.

Comment (jmillikin) [reply rust/platform/channel/BUILD:7]: This is a target name, it's referenced as `@io_bazel_rules_rust//rust/platform/channel` and the constraint values are referenced like `@io_bazel_rules_rust//rust/platform/channel:nightly`.

**Comment by jmillikin** [comment]
> This could use an example. Is CI setup to run 4.0 yet?

For an example it might be better to use toolchains, since those are better known and have been around for a while.

Given the following BUILD targets:
```starlark
platform(
    name = "linux-armv7-stable",
    constraint_values = [
        "@platforms//os:linux",
        "@platforms//cpu:arm",
        "@io_bazel_rules_rust//rust/platform/channel:stable",
    ],
)

platform(
    name = "linux-armv7-nightly",
    constraint_values = [
        "@platforms//os:linux",
        "@platforms//cpu:arm",
        "@io_bazel_rules_rust//rust/platform/channel:nightly",
    ],
)


toolchain(
    name = "rust_darwin-to-linux-armv7-stable",
    exec_compatible_with = [
        "@platforms//cpu:x86_64",
        "@platforms//os:macos",
    ],
    target_compatible_with = [
        "@platforms//cpu:arm",
        "@platforms//os:linux",
        "@io_bazel_rules_rust//rust/platform/channel:stable",
    ],
    toolchain = "@rust_darwin-to-linux-armv7-stable//:toolchain_for_arm-unknown-linux-musleabihf_impl",
    toolchain_type = "@io_bazel_rules_rust//rust:toolchain",
)

toolchain(
    name = "rust_darwin-to-linux-armv7-nightly",
    exec_compatible_with = [
        "@platforms//cpu:x86_64",
        "@platforms//os:macos",
    ],
    target_compatible_with = [
        "@platforms//cpu:arm",
        "@platforms//os:linux",
        "@io_bazel_rules_rust//rust/platform/channel:nightly",
    ],
    toolchain = "@rust_darwin-to-linux-armv7-nightly//:toolchain_for_arm-unknown-linux-musleabihf_impl",
    toolchain_type = "@io_bazel_rules_rust//rust:toolchain",
)
```

I can use now select whether I want to build with the stable or nightly toolchain using the `--platforms` flag, similar to Cargo's `+stable` / `+nightly` feature.

**Comment by hlopko** [comment]
I'll go ahead and merge this PR, IMHO it's useful and uncontroversial. Thank you @jmillikin!

**Pull request merged**